\documentclass{article}
\usepackage[utf8]{inputenc}
\usepackage{amsfonts}
\usepackage{amsmath}
\usepackage{natbib}
\usepackage{graphicx}
\usepackage{array,booktabs,tabularx}
\usepackage{epstopdf}
\usepackage{colortbl, xcolor}
\usepackage{tabularx}
\usepackage{multirow}

\title{Prediction of malarial signal peptides using signalHsmm}
\author{Micha\l{} Burdukiewicz}

\begin{document}

\maketitle
\section{Introduction}

% https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2588515/
% https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2408606/

<<knitrIntro, echo = FALSE, message=FALSE,warning=FALSE>>=
source("my_ggplot_theme.R")

g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

library(dplyr)
library(reshape2)
library(xtable)
library(biogram)

load("./dat/signalhsmm_poster.RData")

freq_nondeg <- read.csv2("./dat/freq_nondeg.csv")[, -1]
freq_deg <- read.csv2("./dat/freq_deg.csv")[, -1]

do_pca <- function(x) 
  x %>% 
  select(-type, -taxon) %>% 
  prcomp(center = TRUE, scale = TRUE) %>% 
  getElement("x") %>% 
  data.frame() %>% 
  select(1, 2) %>% 
  cbind(select(x, type, taxon), .) %>% 
  mutate(type_nice = factor(type, labels = c("Mature peptide", "Signal peptide")),
         taxon_nice = factor(taxon, labels = c("other eukaryotes", "Plasmodium"))) %>% 
  mutate(both = paste0(type_nice, "\n(", taxon_nice, ")\n")) %>% 
  filter(type_nice == "Signal peptide")

dat_deg <- do_pca(freq_deg) 
dat_nondeg <- do_pca(freq_nondeg)

# dat_deg <- read.table("PCAgr.txt", header = TRUE, sep = "\t")
# dat_nondeg <- read.table("PCA.txt", header = TRUE, sep = "\t")
# colnames(dat_deg) <- c("both", "PC1", "PC2")
# colnames(dat_nondeg) <- c("both", "PC1", "PC2")

plot_pca <- function(x)
  ggplot(x, aes(x = PC1, y = PC2, fill = both)) + 
  geom_density_2d(color = "black", contour = TRUE) +
  #geom_point() +
  stat_density2d(aes(fill=both,alpha=..level..), color = "black", contour = TRUE, geom="polygon") +
  scale_linetype_discrete("") +
  scale_fill_manual("", values = c("dodgerblue1", "firebrick1"), 
                    labels = c("Other eukaryotes", "Plasmodiidae")) +
  scale_shape_discrete("") +
  scale_color_discrete("") +
  scale_x_continuous("Principal Component 1\n") +
  scale_y_continuous("Principal Component 2") +
  scale_alpha_continuous(range = c(0.25, 0.4)) +
  guides(alpha = FALSE) +
  my_theme

p1 <- plot_pca(dat_nondeg) 
p2 <- plot_pca(dat_deg)
leg <- g_legend(p1)
@

Heavy adenine-thymine bias of malarial genomes alters amino acid composition of malarial proteins, including signal peptides. Simple PCA analysis of amino acid frequency shows that signal peptides of \textit{Plasmodiidae} do not group with signal peptides of other eukaryotes (Figure \ref{fig:PCA}A).

%signal peptides making them hard to predict using software trained on other eukaryotes.

%Medically significant have unique composition of signal peptides 

\begin{figure}
<<echo = FALSE,message=FALSE,results='asis',fig.width=14, fig.height=13>>=
grid.arrange(textGrob("A", x = 0.75, y = 0.9, gp=gpar(fontsize=32)), 
             p1 + guides(fill = FALSE) + ggtitle("Full alphabet"), 
             textGrob("B", x = 0.75, y = 0.9, gp=gpar(fontsize=32)),
             p2 + guides(fill = FALSE) + ggtitle("Reduced alphabet") , 
             rectGrob(x = unit(0.5, "npc"), y = unit(0.5, "npc"), gp = gpar(col = "white")), leg, 
             nrow = 3, heights = c(0.48, 0.48, 0.04), widths = c(0.04, 0.96))
@ 
\caption{Principal component analysis of amino acid frequency in signal peptides belonging to \textit{Plasmodiidae} and other eukaryotes. A) Frequency of amino acids. B) Frequency of amino acids encoded using the reduced alphabet.}
\label{fig:PCA}
\end{figure}

\section{Reduced alphabet}

We generated 96 reduced amino acid alphabets using combination of physicochemical properties relevant to signal peptide architecture (charge, polarity, hydrophobicity). To assess if reduced amino acid alphabets create more general model of signal peptides, we build a signal peptide predictor (based on hidden semi-Markov models) separately for each alphabet. In a cross-validation experiment (using only eukaryotic proteins) we find a reduced amino acid alphabet providing the best sensitivity (and second best AUC)~(Table~\ref{tab:best_enc}). 

It is important to note, that reduced amino acid alphabet groups together signal peptides belonging to \textit{Plasmodiidae} and other eukaryotes (Figure \ref{fig:PCA}A).

\begin{table}[ht]
\centering
\caption{The best performing reduced amino acid alphabet.}
\begin{tabular}{ll}
  \toprule
Group & Amino acids \\ 
  \midrule
I & D, E, H, K, N, Q, R \\ 
   \rowcolor[gray]{0.85}II & G, P, S, T, Y \\ 
  III & F, I, L, M, V, W \\ 
   \rowcolor[gray]{0.85} IV & A, C \\ 
   \bottomrule
\end{tabular}
\label{tab:best_enc}
\end{table}

\section{Benchmark}

To create benchmark data set, we extracted proteins with signal peptide belonging to members of \textit{Plasmodiidae} (51 proteins after 50\% homology reduction). As negative data set we used proteins without signal peptide from the same taxon (211 proteins after 50\% homology reduction).

As predictor we used \textit{signalHsmm}-2010, hidden semi-Markov model trained on data set of 3,676 eukaryotic proteins with signal peptides added before year 2010 and encoded using the best sensitivity reduced alphabet. signalHsmm was compared to other predictors of signal peptides. As a negative control, we also benchmarked an iteration of signalHsmm that does not employ reduced amino acid alphabet.

\small
\begin{table}[ht]
\centering
\caption{Results of benchmark. Full alphabet: no amino alphabet reduction. hom. 50\%: 50\% homology reduction in the learning data set.
}
%\begin{tabularx}{\textwidth}{@{} Xccc @{}}
\begin{tabularx}{\textwidth}{Xlccc}

  \toprule
 & Sensitivity & Specificity & MCC & AUC \\ 
  \midrule
signalP 4.1 (no tm) \newline \citep{petersen_signalp_2011} & 0.8235 & 0.9100 & 0.6872 & 0.8667 \\ 
   \rowcolor[gray]{0.85}signalP 4.1 (tm) \newline \citep{petersen_signalp_2011} & 0.6471 & 0.9431 & 0.6196 & 0.7951 \\ 
  signalP 3.0 (NN) \newline \citep{DyrlovBendtsen2004783} & 0.8824 & 0.9052 & 0.7220 & 0.8938 \\ 
   \rowcolor[gray]{0.85}signalP 3.0 (HMM) \newline \citep{DyrlovBendtsen2004783} & 0.6275 & 0.9194 & 0.5553 & 0.7734 \\ 
  PrediSi \newline \citep{hiller_predisi:_2004} & 0.3333 & \textbf{0.9573} & 0.3849 & 0.6453 \\ 
   \rowcolor[gray]{0.85}Philius \newline \citep{reynolds_transmembrane_2008} & 0.6078 & 0.9336 & 0.5684 & 0.7707 \\ 
  Phobius \newline \citep{kall_combined_2004} & 0.6471 & 0.9289 & 0.5895 & 0.7880 \\ 
   \hline
   \hline
   \rowcolor[gray]{0.85}signalHsmm-2010 & 0.9804 & 0.8720 & 0.7409 & 0.9262 \\ 
   signalHsmm-2010 (full alphabet) & 0.8431 & 0.9005 & 0.6853 & 0.8718 \\ 
   \bottomrule
\end{tabularx}
\end{table}

\section{Workplan}

Aim: improve performance of signalP for atypical signal peptides using a reduced amino acid alphabet.

\begin{enumerate}
\item Adjust signalP 4.1 for reduced alphabets. I don't have an access to signalP 4.1 source code, but it is possible, that it has hardcoded alphabet of 20 amino acids, so a bit of programming might be necessary.
%\item Generate reduced amino acid alphabets using physicochemical properties appropriate for signal peptides.
%\item Through cross-validation experiment identify the reduced alphabet that maximizes performance measures for sets of unique signal peptides.
\item Benchmark modified signalP 4.1 on external data set(s) of atypical signal peptides and compare with normal signalP 4.1 and signalP 3.0.
\end{enumerate}

\section{Concerns}

It is possible that reduction of the alphabet may improve the general performance of signal peptide prediction, but lower accuracy of cleavage sites prediction. Cleavage sites seem to require more defined motifs than whole signal peptide and may need larger alphabets, possibly even the full amino acid alphabet.

\bibliographystyle{apalike}
\bibliography{signalhsmm_poster}

\end{document}